### ISSUE TITLE
unexpected partition on disks >= 1TB (atari partitions)

### ISSUE THREAD

**Issue by zzswang** [rook/rook#7940]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**

```
vdb    252:16   0    3T  0 disk 
├─vdb2 252:18   0   48G  0 part 
└─vdb3 252:19   0  6.1M  0 part 
```

**Expected behavior:**

```
vdb    ceph_bluestore
```

**How to reproduce it (minimal and precise):**

1. An empty data disk >= 1TiB
2. Prepare disk, using following script https://github.com/rook/rook/blob/master/Documentation/ceph-teardown.md#zapping-devices
until now there still no child partitions in vdb, looks good.
3. Follow [Quickstart.md](https://github.com/rook/rook/blob/master/Documentation/ceph-quickstart.md)
4. Then there will be 2 OSDs on this disk, on in vdb, another in vdb2 which size is 48GiB.

### Workaround [added by @BlaineEXE]
If users wish to work around this issue, my latest suggestion is to partition any device greater than 875GB with a GPT partition table and create a single partition spanning the whole device. The GPT partition at the beginning of the disk should stop Linux from observing any "phantom" Atari partitions, and users will not have to destroy raw disk OSDs later to be re-created as LVM.

**Environment**:
* OS (e.g. from /etc/os-release): Ubuntu 20.04
* Kernel (e.g. `uname -a`):  Linux master-11 5.4.0-73-generic #82-Ubuntu SMP Wed Apr 14 17:39:42 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
* Rook version (use `rook version` inside of a Rook Pod): chart version 1.6.2
* Storage backend version (e.g. for ceph do `ceph -v`): v15.2.11
* Kubernetes version (use `kubectl version`): k3s+1.19

**Comment by zzswang** [comment]
What I find now, if change the disk init script from 

```
sgdisk --zap-all $DISK
dd if=/dev/zero of="$DISK" bs=1M count=100 oflag=direct,dsync
```

to

```
sgdisk --zap-all --clear --mbrtogpt $DISK
sgdisk --zap-all $DISK
```

48GiB extra parition will gone, everything works fine until now (less than a day)

I don't think it might not the problem of teardown scripts, but there should be something wrong with partition table.

**Comment by sp98** [comment]
are you expecting OSDs on disk (`vdb`)  with child disks?  
Rook does not create OSDs on disks with child devices. [See](https://github.com/rook/rook/blob/fcf634e3b7808e4338e7fd261c047aae4793e99b/pkg/clusterd/disk.go#L99)

**Comment by zzswang** [comment]
No, what I'm expecting is one OSD on vdb of the whole disk.

But after step 3, the child partitions come up. Have no idea.

**Comment by zzswang** [comment]
![image](https://user-images.githubusercontent.com/1524745/118776961-255a4480-b8bb-11eb-8c21-9edd3fc45089.png)

**Comment by sp98** [comment]
not sure about this one. I didn't see any partitions when I tried the tear down steps on a 2.4T HDD. 

```

sh-4.4# lsblk
NAME    MAJ:MIN RM   SIZE RO TYPE MOUNTPOINT
xvda    202:0    0   120G  0 disk 
|-xvda1 202:1    0     1M  0 part 
|-xvda2 202:2    0   127M  0 part 
|-xvda3 202:3    0   384M  0 part /boot
`-xvda4 202:4    0 119.5G  0 part /sysroot
xvdf    202:80   0   2.4T  0 disk 
sh-4.4# DISK="/dev/xvdf"
sh-4.4# sgdisk --zap-all $DISK
Creating new GPT entries.
GPT data structures destroyed! You may now partition the disk using fdisk or
other utilities.
sh-4.4# dd if=/dev/zero of="$DISK" bs=1M count=100 oflag=direct,dsync
100+0 records in
100+0 records out
104857600 bytes (105 MB, 100 MiB) copied, 0.865201 s, 121 MB/s
sh-4.4# lsblk -f 
NAME    FSTYPE LABEL      UUID                                 MOUNTPOINT
xvda                                                           
|-xvda1                                                        
|-xvda2 vfat   EFI-SYSTEM E0E1-433E                            
|-xvda3 ext4   boot       c64c8f5a-092b-4f03-a9b2-4a1d2fa62f8f /boot
`-xvda4 xfs    root       a3205ef1-8830-4532-9f22-06cee2dbf7fa /sysroot
xvdf                                                           
sh-4.4# lsblk
NAME    MAJ:MIN RM   SIZE RO TYPE MOUNTPOINT
xvda    202:0    0   120G  0 disk 
|-xvda1 202:1    0     1M  0 part 
|-xvda2 202:2    0   127M  0 part 
|-xvda3 202:3    0   384M  0 part /boot
`-xvda4 202:4    0 119.5G  0 part /sysroot
xvdf    202:80   0   2.4T  0 disk 
sh-4.4# ```

**Comment by zzswang** [comment]
Thanks for your tests.

It is ok, after teardown scripts. The weird thing is only happened after rook-ceph cluster setup.

When rook-ceph is ready, the wrong partitions will come up. If you didn't see it, just try to allocate some PVCs to make sure prepare step is finished.

**Comment by BlaineEXE** [comment]
This seems like it might be a duplicate of https://github.com/rook/rook/issues/7734

**Comment by BlaineEXE** [comment]
Neither Rook nor the `ceph-volume` tool that Rook uses for preparing disks for OSDs create partitions. From what I could see from #7734 the partitions are not created by Rook's `rook-ceph-osd-prepare-` jobs. 

This means that either (a) Ceph daemons are somehow creating the partitions (unlikely given the extensive testing Ceph receives from Teuthology) or (b) there is a system daemon interfering with Ceph's operation on the machines.

@zzswang you can help us rule out (a) as a possibility by running the following test:
1. Uninstall Rook completely, wipe the disks on nodes, and clear the `/var/lib/rook` directory on nodes
1. Investigate the disks on all nodes using `lsblk` after wiping the nodes (give us this log named `lsblk-TIMESTAMP.log`)
1. Install the Rook operator and a new CephCluster
1. Wait for `rook-ceph-osd-prepare-...` jobs to appear
1. Scale the `rook-ceph-operator` deployment to 0 replicas
1. Wait for the `rook-ceph-osd-prepare-...` jobs to finish
1. There should be no `rook-ceph-osd-ID` deployments created or running
1. Investigate the disks on all nodes using `lsblk` every 20 minutes for 12 hours (give us each log named `lsblk-TIMESTAMP.log`)

If any of the `lsblk` logs from the last step show partitions created, we can be sure that it is not the `ceph-osd` process creating a partition inadvertently and that they are created by a system process.

**Comment by zzswang** [comment]
@BlaineEXE  OK

Currently the machines have been used for deploying project. I'll test it later once have some nodes free.

**Comment by brian-fa** [comment]
I'm seeing this issue as well.  There is a lot more data and context in [this](https://github.com/rook/rook/issues/7940) slack thread.

In my testing, the additional partitions [show up](https://rook-io.slack.com/archives/CK9CF5H2R/p1621003250322400?thread_ts=1619618746.137100&cid=CK9CF5H2R) during the `prepare` step.  I know that doesn't mean that it's the prepare job or `ceph-volume` creating these partitions, but there is definitely a correlation.

I'm running Rook/Ceph in AWS.  Each time I collected data, it was on a brand new EC2 instance with a brand new EBS volume.  And like @zzswang, this only happens with disks >=1TB in size.

Also, I don't think it was stated explicitly above: These extra 48GB partitions getting added to the cluster, results in a broken cluster.  I'm not able to mount PVCs while these 48GB partitions/OSDs are in the cluster.

Thanks for any help on this!  I went to production with Rook/Ceph with 900GB disks to workaround this issue, so I hope we can get to the bottom of this.  Let me know what other data would be helpful to track this down.

**Comment by preslavgerchev** [comment]
I'm seeing this as well. I'm running on two identical nodes

**Environment**:

  * OS (e.g. from /etc/os-release): Ubuntu 18.04 LTS
  * Kernel (e.g. uname -a): Linux node-a 4.15.0-72-generic #81-Ubuntu SMP Tue Nov 26 12:20:02 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
  * Rook version (use rook version inside of a Rook Pod): chart version 1.6.3
  * Storage backend version (e.g. for ceph do ceph -v): v16.2.0
  * Kubernetes version (use kubectl version): 1.19.8

I have been experimenting a lot on these 2 nodes with provisioning and tearing down Ceph and the partitions show up every time.
@BlaineEXE I have done what you suggested and have collected logs, you can find them below. I have also included my ceph cluster config and 3 prepare logs per node - one after the operator stop, one after it resumed and one after the osds were up.

[rook-logs.zip](https://github.com/rook/rook/files/6611509/rook-logs.zip)

Some info about the logs and what I observed in general:
 The cluster config consists of 2 nodes:
       * `node-a` - 3.7TB HDD and a 953GB SSD, mounted under `sdb` and `sdd` respectively
       * `node-b` - 3.7TB HDD and a 447G SSD, mounted under `sdb` and `sdd` respectively

There's an extra 3.7TB HDD per node that I did not include on purpose to see what would happen there (`sdc` on both). 

I ran a cron on each node for well over 24 hours after stopping the operator just after the `prepare` jobs started - at no point did a partition show up. Once the `operator` was scaled up and the `osds` popped up so did the partitions. Note that the smaller SSD (the 447GB one) does **not** have a partition, everything else does. I am in no way that familiar with Ceph but that leads me to believe it indeed has to do with the disk size as already suggested 

To be absolutely certain, I cleaned the disks, removed the cluster and recreated it, after which I stopped the operator again once the `prepare` jobs came up. This time I manually created an OSD deployment (copying and tuning some values) and ran it. The following is the logs from when I manually started an OSD:

```
[root@rook-ceph-osd-0-65597fd8cd-thgdg ceph]# lsblk
NAME   MAJ:MIN RM   SIZE RO TYPE MOUNTPOINT
sda      8:0    0 223.6G  0 disk
|-sda1   8:1    0   512M  0 part
`-sda2   8:2    0 223.1G  0 part /var/lib/ceph/osd/ceph-0
sdb      8:16   0   3.7T  0 disk
sdc      8:32   0   3.7T  0 disk
sdd      8:48   0 447.1G  0 disk
[root@rook-ceph-osd-0-65597fd8cd-thgdg ceph]# ps aux
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root         1  0.0  0.0  11896  2936 ?        Ss   18:31   0:00 /bin/bash -c -- while true; do sleep 30; done;
root         7  0.0  0.0  12132  3316 pts/1    Ss   18:32   0:00 /bin/bash
root       112  0.0  0.0  23028  1400 ?        S    18:39   0:00 /usr/bin/coreutils --coreutils-prog-shebang=sleep /usr/bin/sleep 30
root       113  0.0  0.0  44636  3448 pts/1    R+   18:40   0:00 ps aux
[root@rook-ceph-osd-0-65597fd8cd-thgdg ceph]# ceph-bluestore-tool show-label --dev /dev/sdb
{
    "/dev/sdb": {
        "osd_uuid": "54be2843-c861-4f19-85bc-52b42734ca86",
        "size": 4000787030016,
        "btime": "2021-06-07T18:22:31.296695+0000",
        "description": "main",
        "bfm_blocks": "976754688",
        "bfm_blocks_per_key": "128",
        "bfm_bytes_per_block": "4096",
        "bfm_size": "4000787030016",
        "bluefs": "1",
        "ceph_fsid": "c1aa1e8d-3a69-4941-a6a2-3d7992c4247b",
        "kv_backend": "rocksdb",
        "magic": "ceph osd volume v026",
        "mkfs_done": "yes",
        "osd_key": "AQDlY75gxvk/NxAAIem0f+N1WPiyMZgBURne4g==",
        "ready": "ready",
        "whoami": "0"
    }
}
[root@rook-ceph-osd-0-65597fd8cd-thgdg ceph]# ceph-volume raw activate --device /dev/sdb --no-systemd --no-tmpfs
Running command: /usr/bin/chown -R ceph:ceph /var/lib/ceph/osd/ceph-0
Running command: /usr/bin/ceph-bluestore-tool prime-osd-dir --dev /dev/sdb --path /var/lib/ceph/osd/ceph-0 --no-mon-config
Running command: /usr/bin/chown -R ceph:ceph /dev/sdb
Running command: /usr/bin/ln -s /dev/sdb /var/lib/ceph/osd/ceph-0/block
Running command: /usr/bin/chown -R ceph:ceph /var/lib/ceph/osd/ceph-0
--> ceph-volume raw activate successful for osd ID: 0
[root@rook-ceph-osd-0-65597fd8cd-thgdg ceph]# ceph-osd --id "0" --fsid c1aa1e8d-3a69-4941-a6a2-3d7992c4247b --setuser ceph --setgroup ceph --crush-location="root=default host=<redacted> rack=rack-3" --log-to-stderr=false --err-to-stderr=true --mon-cluster-log-to-stderr=tru
e --default-log-to-file=true --ms-learn-addr-from-peer=false --ms-bind-ipv4=true --ms-bind-ipv6=false
2021-06-07T18:40:33.037+0000 7f2d3162a080 -1 Falling back to public interface
2021-06-07T18:40:35.561+0000 7f2d3162a080 -1 osd.0 0 log_to_monitors {default=true}
[root@rook-ceph-osd-0-65597fd8cd-thgdg ceph]# lsblk
NAME   MAJ:MIN RM   SIZE RO TYPE MOUNTPOINT
sda      8:0    0 223.6G  0 disk
|-sda1   8:1    0   512M  0 part
`-sda2   8:2    0 223.1G  0 part /var/lib/ceph/osd/ceph-0
sdb      8:16   0   3.7T  0 disk
|-sdb2   8:18   0    48G  0 part
`-sdb3   8:19   0   6.1M  0 part	
sdc      8:32   0   3.7T  0 disk
sdd      8:48   0 447.1G  0 disk
[root@rook-ceph-osd-0-65597fd8cd-thgdg ceph]# ps aux
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root         1  0.0  0.0  11896  2936 ?        Ss   18:31   0:00 /bin/bash -c -- while true; do sleep 30; done;
root         7  0.0  0.0  12132  3316 pts/1    Ss   18:32   0:00 /bin/bash
root       131  0.0  0.0  23028  1392 ?        S    18:40   0:00 /usr/bin/coreutils --coreutils-prog-shebang=sleep /usr/bin/sleep 30
ceph       141  1.6  0.0 937948 45560 ?        Ssl  18:40   0:00 ceph-osd --id 0 --fsid c1aa1e8d-3a69-4941-a6a2-3d7992c4247b --setuser ceph --setgroup ceph --crush-location=root=default host=<redacted> rack=rack-3 --log-to-stderr=false --err-to-stderr=true --mon-cluster-lo
root       581  0.0  0.0  44636  3408 pts/1    R+   18:40   0:00 ps aux
```
I do not have much experience with Ceph so I wouldn't know where to look for follow-up but I do hope that the above helps, let me know if there is any other data that might be useful.

**Comment by BlaineEXE** [comment]
@preslavgerchev thanks for the really good info. It's really helpful to learn that the partitions only show up once OSDs begin. I will keep looking into this. I haven't been able to repro yet, which is making this very difficult.

**Comment by BlaineEXE** [comment]
I wonder if there are udev rules on the node set up to trim/discard the raw disks which the base system might think is empty. Are the disks affected mounted to fstab in any way on the host? Are there udev rules to discard disks?

**Comment by dalingng** [comment]
The partition will be displayed when osd is running, but the partition reappears after stopping the osd service

**Comment by Krast76** [comment]
Hi there 👋 

I'm facing the same issue, but with an on-premise deployment. (Obviously i can't reduce the size of my physical disks).

After debugging for days, i think that i've found the root cause.

I saw on my kernel.log this message : 

```bash
[  208.192028]  sdh: AHDI sdh2 sdh3 sdh4
```

And i found on the kernel this : 

```
root@wanda:~/kernel/linux-5.4.0# grep -rni 'AHDI' *
block/partitions/atari.c:43:	int part_fmt = 0; /* 0:unknown, 1:AHDI, 2:ICD/Supra */
block/partitions/atari.c:73:	strlcat(state->pp_buf, " AHDI", PAGE_SIZE);
```

It seems that our partitions are recognized as "ATARI" partition by the kernel.

You can find on your servers/instances if ATARI partitions are supported by the command below : 

```bash
$ cat /boot/config-$(uname -r) | grep -i atari
CONFIG_ATARI_PARTITION=y
```

After recompiling my kernel with this function disabled the bug seems gone 🎉 

I saw this too : https://bugs.launchpad.net/ubuntu/+source/linux-aws/+bug/1908264

That means if you have this issue and you're running on a cloud provider. A kernel with ATARI support disabled will be released soon

On CentOS (7 here) the ATARI support seems disabled : 

```bash
$ cat /boot/config-3.10.0-1160.25.1.el7.x86_64 | grep -i ATARI
# CONFIG_ATARI_PARTITION is not set
```

**Comment by BlaineEXE** [comment]
Thanks @Krast76. This is really helpful. 

ATARI support is going to be removed from Ubuntu for AWS, but there is no indication I could find that it will be removed for non-cloud images.

The launchpad bug mentions that ATARI partitions have "very broad specifications", and this is a little more low-level info into that: https://bug-parted.gnu.narkive.com/UiVUXd5u/bug-31111-partition-table-wrongly-identified-as-atari

Currently, it seems best to me if we are able to head this issue off from the Rook (or Ceph) side of things and ignore ATARI partitions.

I'll do a little digging into the logs people have provided to see if there is a way we can filter those out.

**Comment by BlaineEXE** [comment]
There is a Rook issue that has evidence of an ATARI partition here: https://github.com/rook/rook/issues/2047 which is listed from the `blkid` command.

Some tools (reportedly even versions of `blkid`) don't support ATARI partitions and thus can't recognize them as anything more than "partition" because the Linux kernel recognizes them (ref: https://bugs.launchpad.net/ubuntu/+source/util-linux/+bug/1531404/comments/10)

Given that we might not be able to detect ATARI partitions in Rook, we probably want to treat this as a documentation update, and we should follow up with the Ceph community to make sure they are aware of the possibility also.

We should provide a warning in the documentation that kernels built with ATARI partition support should be avoided.
We have ways to determine this.

I quickly tested on container images for major distributions: ubuntu, centos (7 and 8), fedora, debian, archlinux, alpine:
```sh
$ zcat /proc/config.gz | grep ATARI
# CONFIG_ATARI_PARTITION is not set
```

**Comment by BlaineEXE** [comment]
@Krast76 @dalingng @preslavgerchev @zzswang are any of you able to do a bit of data gathering for me?

From a Rook cluster with the partition issue present, could you `kubectl exec -it` into a `rook-ceph-osd-ID` pod and run some `blkid` (and other) commands for me? Preferably, on a cluster running Ceph v15.2.0 since that would likely have the oldest version of `blkid`.

- `ceph version` 
- `cat /etc/system-release`
- `blkid` just by itself
- `blkid --probe /dev/sdX<part>` (e.g., /dev/sdd1) for extra unexpected partitions

**Comment by BlaineEXE** [comment]
I did some testing on an Ubuntu VM. I can create atari partitions with `parted`, but `lsblk` and `blkid` won't detect the type as atari. `blkid -k` does not list atari as supported using the version in the ceph/ceph:v15.2.0 container. The ceph container does seem to have parted installed, and it is possible to list a block device like below:

```console
# parted -s /dev/sdb1 print
Model: Unknown (unknown)
Disk /dev/sdb1: 1509kB
Sector size (logical/physical): 512B/512B
Partition Table: atari
Disk Flags:

Number  Start  End  Size  Type  File system  Flags

```

**Comment by BlaineEXE** [comment]
@Krast76 @dalingng @preslavgerchev @zzswang @linkvt @brian-fa @dalingng 

Thank you all for your patience with and help uncovering this issue. It has been really bizarre. I believe #8200 will solve the issue, and we will backport it to have it released with Rook v1.6.7. I'll keep you updated here once we have a test build for you to try and hopefully verify for us since we've been having trouble reproducing the issue ourselves.

**Comment by linkvt** [comment]
Hi @BlaineEXE,  thanks for the update and thanks to the others here for finding the issue!

I have only one question that I didn't see answered in the PR, what should I do with my existing OSDs right now, should I scratch them and reinstall them with rook 1.6.7 running? I setup a device filter to exclude partitions which prevented the additional partitions of this bug, but not sure if my existing OSDs are now good to go for the future.

**Comment by BlaineEXE** [comment]
> Hi @BlaineEXE, thanks for the update and thanks to the others here for finding the issue!
> 
> I have only one question that I didn't see answered in the PR, what should I do with my existing OSDs right now, should I scratch them and reinstall them with rook 1.6.7 running? I setup a device filter to exclude partitions which prevented the additional partitions of this bug, but not sure if my existing OSDs are now good to go for the future.

If you are using that workaround, you should not need to scratch the OSDs. Only if Rook has created OSDs on both the main disk/partition and one of the "phantom" partitions will you need to scratch those OSDs.

**Comment by BlaineEXE** [comment]
Backported to 1.6 in #8223

**Comment by BlaineEXE** [comment]
@Krast76 @dalingng @preslavgerchev @zzswang @linkvt @brian-fa @dalingng

It would be good to verify the fix (now built into the rook/ceph:v1.6.0.508.g037945a image) is working on the real-world error. I did my best to test this using manually-created atari partitions, but things can always slip through.

**Comment by zzswang** [comment]
@BlaineEXE 

Thank you for fixing it.  I'm walk away for a while since some busy work.

I will test it as soon as possible.

**Comment by preslavgerchev** [comment]
@BlaineEXE  I just tried a fresh deployment using rook/ceph:v1.6.0.508.g037945a. This time I set `useAllNodes` and `useAllDevices` to `true`. OSDs still got deployed on these freshly created partitions, however upon inspecting them with `parted` I get the following output (this is on an `Ubuntu 20.04.2`):
```sh
root@node-a:~# parted --script --machine /dev/sdb print
Error: /dev/sdb: unrecognised disk label
BYT;
/dev/sdb:4001GB:scsi:512:512:unknown:ATA ST4000NM0035-1V4:;
root@node-a:~# parted --script --machine /dev/sdb2 print
Error: /dev/sdb2: unrecognised disk label
BYT;
/dev/sdb2:51.5GB:unknown:512:512:unknown:Unknown:;
root@node-a:~# parted --script --machine /dev/sdb3 print
Error: /dev/sdb3: unrecognised disk label
BYT;
/dev/sdb3:6458kB:unknown:512:512:unknown:Unknown:;

```

So in my case at least I would say it's expected that OSDs are deployed there since it's not an `atari` partition and is not skipped therefore.

However, there seems to be another issue with this: If the 'parent' OSD (the one deployed onto `sdb`) goes down or needs to be redeployed for whatever reason, the operator will not pick it up anymore during the prepare job. That's due to the hardware discovery step that skips all blocks and/or partitions that have children, see some bits of  logs below:

```
2021-07-02 09:34:05.862463 I | inventory: skipping device "sdb" because it has child, considering the child instead.
2021-07-02 09:34:08.013700 D | exec: Running command: stdbuf -oL ceph-volume --log-path /tmp/ceph-log raw list --format json
2021-07-02 09:34:08.637147 D | cephosd: {
"11": {
"ceph_fsid": "b08e48d4-a25c-4953-a16f-14a102af3c46",
"device": "/dev/sdb2",
"osd_id": 11,
"osd_uuid": "8e7528b9-fcfa-461d-b041-f4ea5d570a0a",
"type": "bluestore"
},
"14": {
"ceph_fsid": "b08e48d4-a25c-4953-a16f-14a102af3c46",
"device": "/dev/sdc2",
"osd_id": 14,
"osd_uuid": "19e83bad-9cb0-48bf-aae2-30cdbe122f6e",
"type": "bluestore"
},
"6": {
"ceph_fsid": "b08e48d4-a25c-4953-a16f-14a102af3c46",
"device": "/dev/sdd",
"osd_id": 6,
"osd_uuid": "4809f21e-4cf8-4c6c-8ef0-84d37b5c97ee",
"type": "bluestore"
}
}
2021-07-02 09:34:08.637368 D | exec: Running command: lsblk /dev/sdb2 --bytes --nodeps --pairs --paths --output SIZE,ROTA,RO,TYPE,PKNAME,NAME,KNAME
2021-07-02 09:34:08.643348 I | cephosd: setting device class "hdd" for device "/dev/sdb2"
2021-07-02 09:34:08.643401 D | exec: Running command: lsblk /dev/sdc2 --bytes --nodeps --pairs --paths --output SIZE,ROTA,RO,TYPE,PKNAME,NAME,KNAME
2021-07-02 09:34:08.648956 I | cephosd: setting device class "hdd" for device "/dev/sdc2"
2021-07-02 09:34:08.649031 D | exec: Running command: lsblk /dev/sdd --bytes --nodeps --pairs --paths --output SIZE,ROTA,RO,TYPE,PKNAME,NAME,KNAME
2021-07-02 09:34:08.653255 D | exec: Running command: sgdisk --print /dev/sdd
2021-07-02 09:34:08.657963 I | cephosd: setting device class "ssd" for device "/dev/sdd"
2021-07-02 09:34:08.658015 I | cephosd: 3 ceph-volume raw osd devices configured on this node
2021-07-02 09:34:08.658108 I | cephosd: devices = [{ID:11 Cluster:ceph UUID:8e7528b9-fcfa-461d-b041-f4ea5d570a0a DevicePartUUID: DeviceClass:hdd BlockPath:/dev/sdb2 MetadataPath: WalPath: SkipLVRelease:true Location:root=default host=zeuclu10 LVBackedPV:false CVMode:raw Store:bluestore TopologyAffinity:} {ID:14 Cluster:ceph UUID:19e83bad-9cb0-48bf-aae2-30cdbe122f6e DevicePartUUID: DeviceClass:hdd BlockPath:/dev/sdc2 MetadataPath: WalPath: SkipLVRelease:true Location:root=default host=node-a LVBackedPV:false CVMode:raw Store:bluestore TopologyAffinity:} {ID:6 Cluster:ceph UUID:4809f21e-4cf8-4c6c-8ef0-84d37b5c97ee DevicePartUUID: DeviceClass:ssd BlockPath:/dev/sdd MetadataPath: WalPath: SkipLVRelease:true Location:root=default host=zeuclu10 LVBackedPV:false CVMode:raw Store:bluestore TopologyAffinity:}]


I think all of this summed up goes as following:
1. Initial `prepare` discovers the 'normal' blocks - `sdb`, `sdc`, `sdd`.
2. OSDs are deployed there which also results in `sdb2`, `sdb3`, etc.
3. Next time prepare runs it skips the `sdb`, `sdc` blocks as they have children and now their partitions are considered (only `sdb2` as `sdb3` is too small)
4. OSDs are now deployed there too
5. If the `sdb` OSD ever goes down, the operator will never bring it back up.

Shall I open a new issue for this @BlaineEXE ?

**Comment by BlaineEXE** [comment]
Thanks for the details @preslavgerchev. It seems this rabbit hole goes deeper, and this issue is not fixed. I'm reopening it.

My best guess for why this is happening is that parted has a stricter definition for Atari partitions than the kernel and therefore registers the partitions as `unknown` after the kernel sees them as Atari.

I have started working on #8200 which will change Rook's behavior to only deploy OSDs on empty GPT partitions to avoid this issue and a whole class of related issues. It will unfortunately mean that users who want to use partitions will have to prepare their systems differently, but we think this will be worth it to fix this bug.

While I continue working on the new fix, the existing workaround to use `deviceFilter: ^sd[a-z]+` to target only disks should still work.

**Comment by Mutantt** [comment]
Hi 
I have the same problem and also used  deviceFilter: ^sd[a-z]+ but nothing changed.every time I deploy new cluster between 17 ~ 22 OSDs appear (we have 12 raw disk with 1.7 TB each). when I use ceph osd tree command the result is 12 OSDs are up and others are down.

**Comment by hbahadorzadeh** [comment]
We have the same problem and I've checked you code, you're using parted command to check disk partition table type which definitely does not recognize Atari file system(I've report this bug to parted team). Instead you cloud use blkid which give you a correct output.

```
$ parted  --machine --script /dev/sdx print
BYT;
/dev/sdx:1800GB:scsi:512:4096:unknown:xxxxx:;
$ sudo blkid /dev/sdx
/dev/sdx: TYPE="ceph_bluestore" PTTYPE="atari"
```
I hope this helps you resolve the issue.

**Comment by sethjones** [comment]
In order to work around this a bit, is there a config setting for forcing rook to use the legacy lvm method for formatting new disk?

**Comment by travisn** [comment]
> In order to work around this a bit, is there a config setting for forcing rook to use the legacy lvm method for formatting new disk?

@sethjones There is not currently. See [this comment](https://github.com/rook/rook/issues/7940#issuecomment-873106778) for a deviceFilter workaround

**Comment by BlaineEXE** [comment]
> We have the same problem and I've checked your code, you're using parted command to check disk partition table type which definitely does not recognize Atari file system(I've report this bug to parted team). Instead you cloud use blkid which give you a correct output.
> 
> ```
> $ parted  --machine --script /dev/sdx print
> BYT;
> /dev/sdx:1800GB:scsi:512:4096:unknown:xxxxx:;
> $ sudo blkid /dev/sdx
> /dev/sdx: TYPE="ceph_bluestore" PTTYPE="atari"
> ```
> 
> I hope this helps you resolve the issue.

Not all versions of blkid support atari. 

On Ubuntu 18.04, `blkid from util-linux 2.31.1  (libblkid 2.31.1, 19-Dec-2017)` does not support atari.
But on Centos 8, `blkid from util-linux 2.32.1  (libblkid 2.32.1, 16-Jul-2018)` does support atari.

**Comment by BlaineEXE** [comment]
I looked at the linux code ([header](https://github.com/torvalds/linux/blob/master/block/partitions/atari.h) & [code](https://github.com/torvalds/linux/blob/master/block/partitions/atari.c)) as well as the [Atari AHDI disk spec](https://info-coach.fr/atari/documents/_mydoc/Atari_HD_File_Sytem_Reference_Guide.pdf) (see PART II) and tracked down the source of our woes with a hexdump from an affected disk. The analysis is posted below.

[Atari-Analysis.pdf](https://github.com/rook/rook/files/6812610/Atari-Analysis.pdf)

**Comment by Mutantt** [comment]
> > In order to work around this a bit, is there a config setting for forcing rook to use the legacy lvm method for formatting new disk?
> 
> @sethjones There is not currently. See [this comment](https://github.com/rook/rook/issues/7940#issuecomment-873106778) for a deviceFilter workaround

@travisn Unfortunately this doesn't work, I used  deviceFilter: ^sd[a-z]$ instead of deviceFilter: ^sd[a-z]+ and I got the number of OSDs that I expected to appear in cluster,but I faced with new problem. version detect pods appear in cluster every minute and prepared osd pods running again and again.is it ok ?

**Comment by sethjones** [comment]
I can also confirm that the work around did not work for me as well.  I have a recent disk that was added to an unused node, running v1.6.7 with the device filter, and still have multiple partitions being created.  But it sounds like a fix might be on the horizon.

**Comment by BlaineEXE** [comment]
> I can also confirm that the work around did not work for me as well. I have a recent disk that was added to an unused node, running v1.6.7 with the device filter, and still have multiple partitions being created. But it sounds like a fix might be on the horizon.

The workaround will not stop the extra partitions from being created. That is due to a series of bugs in the Linux kernel when Atari support is enabled. The workaround should stop Rook from running OSDs on the extra partitions, however, which will prevent corruption.

**Comment by BlaineEXE** [comment]
RE: @Mutantt 

> > > In order to work around this a bit, is there a config setting for forcing rook to use the legacy lvm method for formatting new disk?
> > 
> > 
> > @sethjones There is not currently. See [this comment](https://github.com/rook/rook/issues/7940#issuecomment-873106778) for a deviceFilter workaround
> 
> @travisn Unfortunately this doesn't work, I used deviceFilter: ^sd[a-z]$ instead of deviceFilter: ^sd[a-z]+ and I got the number of OSDs that I expected to appear in cluster,but I faced with new problem. version detect pod appear in cluster every minute and prepared osd pods run again and again.is it ok ?

This sounds like the operator is having some sort of reconcile issue. This might be a bug, but it would be a separate one. If you like, you can open up a new issue. Please make sure to include the operator logs in with debug log level enabled.

**Comment by BlaineEXE** [comment]
Sorry @Mutantt and @sethjones. I think the correct workaround should be `deviceFilter: ^sd[a-z]+$` to ensure it stops matching after the alpha characters. I updated the workaround recommendations posted previously.

**Comment by BlaineEXE** [comment]
Update on this: 3 of the fixes I have tried have failed. Even if Rook doesn't allow Ceph to provision the "phantom" partitions with OSDs, `ceph-volume` still gets confused and reports only the phantom partitions in `ceph-volume raw inventory`. These phantom partitions are almost always 48GB in size (smaller than the actual disk), and they have malformed information. This will be a problem if any of the users disks change names (e.g., /dev/sda becomes /dev/sdc after a reboot), because the init container will not be able to recover and find the original disk. 

This also means that any users who have been affected by the "phantom" Atari partition behavior should destroy and re-create their OSDs once the fix is in place in v1.6.8 so they get stable LVM OSDs back. 

The fix I will implement will be to disable raw OSD provisioning on disks. Raw provisioning will still happen on partitions since partitions are not affected by this bug.

If users wish to work around this issue, my latest suggestion is to partition any device greater than 875GB with a GPT partition table and create a single partition spanning the whole device. The GPT partition at the beginning of the disk should stop Linux from observing any "phantom" Atari partitions, and users will not have to destroy raw disk OSDs later to be re-created as LVM.

@Krast76 @dalingng @preslavgerchev @zzswang @sethjones @Mutantt @hbahadorzadeh @linkvt please see my latest workaround advice above. Thank you all for your patience while I've been trying to mitigate this bug.

**Comment by sethjones** [comment]
> Update on this: 3 of the fixes I have tried have failed. Even if Rook doesn't allow Ceph to provision the "phantom" partitions with OSDs, `ceph-volume` still gets confused and reports only the phantom partitions in `ceph-volume raw inventory`. These phantom partitions are almost always 48GB in size (smaller than the actual disk), and they have malformed information. This will be a problem if any of the users disks change names (e.g., /dev/sda becomes /dev/sdc after a reboot), because the init container will not be able to recover and find the original disk. 
> 
> This also means that any users who have been affected by the "phantom" Atari partition behavior should destroy and re-create their OSDs once the fix is in place in v1.6.8 so they get stable LVM OSDs back. 
> 
> The fix I will implement will be to disable raw OSD provisioning on disks. Raw provisioning will still happen on partitions since partitions are not affected by this bug.
> 
> If users wish to work around this issue, my latest suggestion is to partition any device greater than 875GB with a GPT partition table and create a single partition spanning the whole device. The GPT partition at the beginning of the disk should stop Linux from observing any "phantom" Atari partitions, and users will not have to destroy raw disk OSDs later to be re-created as LVM.
> 
> @Krast76 @dalingng @preslavgerchev @zzswang @sethjones @Mutantt @hbahadorzadeh @linkvt please see my latest workaround advice above. Thank you all for your patience while I've been trying to mitigate this bug.

So the new behavior in 1.6.8 will revert to the previous for the time being?  Rook will detect empty disks and create an LVM?

**Comment by BlaineEXE** [comment]
> So the new behavior in 1.6.8 will revert to the previous for the time being? Rook will detect empty disks and create an LVM?

@sethjones that is the plan, partially. Raw mode will still be used for partitions. But the old behavior will be reverted for whole disks.

**Comment by BlaineEXE** [comment]
In order to get back to using raw mode, there are a couple of things we need to fix. A user found a raw mode implementation issue in #8316 recently for one.

Related to the Atari partition issue:  `ceph-volume raw inventory` reports the phantom partitions (with malformed OSD information) **instead** of the originally-created disk-based OSDs. If disks change their friendly names (e.g., 'sda' becomes 'sdb' after a system reboot), `ceph-volume`'s behavior can cause OSDs to fail during Pod initialization when Pods are restarted.

I spoke to Sage Weil from the Ceph project today, and the best holistic strategy currently seems to be to fix the issue in `ceph-volume` code, then re-enable raw mode in Rook for versions of Ceph with the fix. Re-enabling raw mode will also need to address #8316 as well.
